### PR TITLE
feat: amplifier deployment scripts

### PR DESCRIPTION
**Pull request by cjcobb23**
https://axelarnetwork.atlassian.net/browse/AXE-2740

### CODE REVIEW

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:13]:
```suggestion
const getInstantiateMsg = async (contractName, config, chain) => {
```

File: cosmwasm/utils.js
Comment (jcs47) [inline cosmwasm/utils.js:12]:
```suggestion
const uploadContract = async (client, wallet, config, options) => {
```

I don't feel strongly about it, but it feels more natural like this to me.

File: cosmwasm/utils.js
Comment (jcs47) [inline cosmwasm/utils.js:18]: Should this value really be hardcoded?

File: cosmwasm/utils.js
Comment (jcs47) [inline cosmwasm/utils.js:32]:
```suggestion
const instantiateContract = async (client, wallet, contractName, initMsg, config, options) => {
```

File: cosmwasm/utils.js
Comment (jcs47) [inline cosmwasm/utils.js:37]: same question w.r.t. harcoded values

File: cosmwasm/utils.js
Comment (jcs47) [inline cosmwasm/utils.js:59]: this function doesn't seem to be used outside of this package

File: cosmwasm/deploy-contract.js
Comment (jcs47) [inline cosmwasm/deploy-contract.js:14]: isn't there a function to specifically validate cosmwasm addresses somewhere in `cosmjs` ? same question for the rest of the functions

File: cosmwasm/deploy-contract.js
Comment (jcs47) [inline cosmwasm/deploy-contract.js:217]:
```suggestion
const makeInstantiateMsg = (contractName, chain, config) => {
```

File: cosmwasm/deploy-contract.js
Comment (jcs47) [inline cosmwasm/deploy-contract.js:330]: why not fetch the contract name from `options` within `instantiateContract`

File: cosmwasm/deploy-contract.js
Comment (jcs47) [inline cosmwasm/deploy-contract.js:299]: This function doesn't seem to put the script on hold waiting for user confirmation, like it is done for the equivalent script dedicated to evm smart contracts

File: cosmwasm/deploy-contract.js
Comment (jcs47) [inline cosmwasm/deploy-contract.js:380]: if it is decided that this script should wait for user confirmation, than you should also add an `-y/--yes` flag to override confirmation. Particular useful for devnet deployment.

**Review by jcs47 — CHANGES_REQUESTED** [review]

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:321]:
```suggestion
        if (result.address) {
```

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:305]: Same here.

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:330]:
```suggestion
    if (!options.uploadOnly) {
```

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:358]:
```suggestion
    const undefinedChain = chains.find((chain) => !config.chains[chain.toLowerCase()] && chain !== 'none');

    if (undefinedChain) {
        throw new Error(`Chain ${undefinedChain} is not defined in the info file`);
    }
```
js is actually a very functional language.

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:365]:
```suggestion
    fawait Promise.all(
        chains.map((chain) =>
            deploy(options, config.chains[chain.toLowerCase()], config).then(() => {
                saveConfig(config, options.env);

                options.reuseCodeId = true;
            }),
        ),
    );
```
Can the deployment for each chain be "paralleled"? If so, this would have a much better performance. JS is indeed run by a single-thread, but the event loop is pretty powerful at switching back-and-forth between contexts.

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:13]:
```suggestion
    const [account] = await wallet.getAccounts();
    const { artifactPath, contractName, instantiate2, salt, contractName } = options;
```
I'd definitely recommend deconstructing. It's a lot easier to debug when some field is missing cuz you wouldn't have to go through the entire function to tell which are being used.

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:12]:
```suggestion
const uploadContract = async (config, { artifactPath, contractName, instantiate2, salt, contractName }, wallet, client) => {
```
This would work, too. I'll leave it to you.

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:39]: If `let` is sometimes OK, `var` is an absolutely no in js.

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:55]:
```suggestion
    return (
        options.instantiate2
            ? client.instantiate2(
                  account.address,
                  contractConfig.codeId,
                  new Uint8Array(Buffer.from(getSaltFromKey(options.salt || options.contractName).slice(2), 'hex')),
                  initMsg,
                  contractName,
                  initFee,
              )
            : client.instantiate(account.address, contractConfig.codeId, initMsg, contractName, initFee)
    ).then(({ contractAddress }) => contractAddress);
```
JS supports ternary operators! Use it.

File: evm/utils.js
Comment (fish-sammy) [inline evm/utils.js:215]:
```suggestion
const isStringArray = (arr) => Array.isArray(arr) && arr.every(isString);
```
This should be a one-liner.

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:20]: No var.

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:10]:
```suggestion
const pascalToSnake = (str) => str.replace(/([A-Z])/g, (group) => `_${group.toLowerCase()}`).replace(/^_/, '');
```

File: cosmwasm/utils.js
Comment (cjcobb23) [reply cosmwasm/utils.js]: good catch, changed it to read from the config

File: cosmwasm/deploy-contract.js
Comment (jcs47) [reply cosmwasm/deploy-contract.js:387]: @cjcobb23 address this before I approve

**Review by jcs47 — APPROVED** [review]

File: cosmwasm/deploy-contract.js
Comment (cjcobb23) [reply cosmwasm/deploy-contract.js]: yes it's a boolean, it's true if the user wants to reuse the codeId stored in the config.

File: cosmwasm/deploy-contract.js
Comment (cjcobb23) [reply cosmwasm/deploy-contract.js]: chained them as best I could

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [reply cosmwasm/deploy-contract.js]: ah ok fair enough

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:47]:
```suggestion

    return client.upload(account.address, wasm, uploadFee).then(({ checksum, codeId }) => {
        const address = instantiate2
            ? instantiate2Address(
                  fromHex(checksum),
                  account.address,
                  fromHex(getSaltFromKey(salt || contractName.concat(chainNames))),
                  'axelar',
              )
            : null;

        return { codeId, address };
    });
```

File: cosmwasm/utils.js
Comment (fish-sammy) [inline cosmwasm/utils.js:71]:
```suggestion
const instantiateContract = (client, wallet, initMsg, config, { contractName, salt, instantiate2, chainNames }) =>
    wallet
        .getAccounts()
        .then(([account]) => {
            const contractConfig = config.axelar.contracts[contractName];
            const {
                axelar: { gasPrice, gasLimit },
            } = config;
            const initFee = calculateFee(gasLimit, GasPrice.fromString(gasPrice));

            return instantiate2
                ? client.instantiate2(
                      account.address,
                      contractConfig.codeId,
                      fromHex(getSaltFromKey(salt || contractName.concat(chainNames))),
                      initMsg,
                      contractName,
                      initFee,
                  )
                : client.instantiate(account.address, contractConfig.codeId, initMsg, contractName, initFee);
        })
        .then(({ contractAddress }) => contractAddress);
```

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:16]:
```suggestion
const validateAddress = (address) => isString(address) && isValidCosmosAddress(address);
```

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:307]:
```suggestion
const prepareWallet = ({ mnemonic }) => DirectSecp256k1HdWallet.fromMnemonic(mnemonic, { prefix: 'axelar' });
```
Why make it async?

File: cosmwasm/deploy-contract.js
Comment (fish-sammy) [inline cosmwasm/deploy-contract.js:4]:
```suggestion
const { isNil, ... } = require('lodash');
```

**Review by fish-sammy — APPROVED** [review]

**Review by milapsheth — APPROVED** [review]

**Pull request merged**
